docs/content/en/docs/Design/relinearization_ilp.md

```diff
@@ -0,0 +1,199 @@
+---
+title: Optimizing relinearization
+weight: 10
+---
+
+This document outlines the integer linear program model used in the
+[`optimize-relinearization`](https://heir.dev/docs/passes/optimizerelinearizationpasses/#-optimize-relinearization)
+pass.
+
+## Background
+
+In vector/arithmetic FHE, RLWE ciphertexts often have the form $\\mathbf{c} =
+(c_0, c_1)$, where the details of how $c_0$ and $c_1$ are computed depend on the
+specific scheme. However, in most of these schemes, the process of decryption
+can be thought of as taking a dot product between the vector $\\mathbf{c}$ and a
+vector $(1, s)$ containing the secret key $s$ (followed by rounding).
+
+In such schemes, the homomorphic multiplication of two ciphertexts $\\mathbf{c}
+= (c_0, c_1)$ and $\\mathbf{d} = (d_0, d_1)$ produces a ciphertext $\\mathbf{f}
+= (f_0, f_1, f_2)$. This triple can be decrypted by taking a dot product with
+$(1, s, s^2)$.
+
+With this in mind, each RLWE ciphertext $\\mathbf{c}$ has an associated _key
+basis_, which is the vector $\\mathbf{s_c}$ whose dot product with $\\mathbf{c}$
+decrypts it.
+
+Usually a larger key basis is undesirable. For one, operations in a higher key
+basis are more expensive. Repeated multiplications exponentially increase the
+length of the key basis. So to avoid this, an operation called _relinearization_
+was designed that converts a ciphertext from a given key basis back to $(1, s)$.
+Doing this requires a set of _relinearization keys_ to be provided by the client
+and stored by the server.
+
+In general, key bases can be arbitrary. Rotation of an RLWE ciphertext by a
+shift of $k$, for example, first applies the automorphism $x \\mapsto x^k$. This
+converts the key basis from $(1, s)$ to $(1, s^k)$, and more generally maps $(1,
+s, s^2, \\dots, s^d) \\mapsto (1, s^k, s^{2k}, \\dots, s^{kd})$. Most FHE
```

Comment:
Do rotations do that, though? There should be a difference between $(1, s(X)^k)$ (what's in the doc but just making the polynomial nature of $s$ explicit)  and $(1, s(X^k))$ (what a rotation by k produces, prior to key-switching)? Unless there's some ring property that makes these two things magically the same?

Reply:
Also pinging @asraa since I guess this is the continuation of the discussion we had in the PR ;)

Reply:
I think you're right and I was being sloppy. Will wait for Asra to confirm.

Reply:
Man someone needs to write a good tutorial about this stuff 😛